Pep 690 v3 (#4)
Kronuz/peps: pep-0690.rst

accidental leakage of lazy objects is to have the dictionary itself be
responsible to ensure resolution of lazy objects on lookup.

When a lookup finds that the key references a lazy object, it resolves the lazy

> object immediately before returning it. To avoid having side effects mutating
> dictionaries midway an iteration, all lazy objects in a dictionary are resolved
> prior starting an iteration; this could incur a performance penalty when using

Reviewer: ```suggestion
object immediately before returning it. To avoid side effects mutating
dictionaries midway through iteration, all lazy objects in a dictionary are resolved
prior to starting an iteration; this could incur a performance penalty when using
```

bulk iterations (``iter(dict)``, ``reversed(dict)``, ``dict.__reversed__()``,
``dict.keys()``, ``iter(dict.keys())`` and ``reversed(dict.keys())``). To avoid
this performance penalty on the vast majority of dictionaries, which never
contain any lazy objects, we steal a bit from the ``dk_kind`` field

> (`dk_lazy_imports` <https://github.com/python/cpython/blob/3.11/Include/internal/pycore_dict.h#L105>`_),

Reviewer: This link line number seems wrong. It's also confusing to link the text `dk_lazy_imports` to a version of the code (3.11 branch) that doesn't have `dk_lazy_imports` in it. You probably don't need to link to the code at all here?
```suggestion
for a new `dk_lazy_imports` bitfield```

> Specification

Reviewer: Commenting here because it won't let me comment on unchanged code. The paragraph above in Rationale that starts with "To avoid a performance penalty..." is not accurate for the current implementation and needs to be updated. For one thing, we no longer use a specialized lookup kind; instead we steal a bit from the kind and track it separately. And secondly, we don't use this bit to avoid extra cost on lookup for most dictionaries. Instead we have decided (although I think we should verify this in benchmarks) that the cost of object type pointer comparison is so low that it doesn't matter.

=============

Lazy imports are opt-in, and they can be:

* Globally enabled, either via a new ``-L`` flag to the Python interpreter or
  via a call to a new ``importlib.set_lazy_imports()`` function, which makes
  *all* new relevant imports after the call immediately lazy.

* Enabled in a specific module, via ``importlib.enable_lazy_imports_in_module()``.

When the flag ``-L`` is passed to the Python interpreter, a new
``sys.flags.lazy_imports`` is set to ``True``, otherwise it exists as ``False``.
This flag is used to propagate ``-L`` to new Python subprocesses.

> The flag in ``sys.flags.lazy_imports``, not necessarily reflects the current

Reviewer: ```suggestion
The flag in ``sys.flags.lazy_imports`` does not necessarily reflect the current
```

Specification
=============

Lazy imports are opt-in, and they can be:

* Globally enabled, either via a new ``-L`` flag to the Python interpreter or
  via a call to a new ``importlib.set_lazy_imports()`` function, which makes

>   *all* new imports after the call immediately lazy.

Reviewer 2: this is not accurate. "*all*" implies, well, "all" :) but in practice it applies to the same imports that would have been lazy under `-L` (e.g. module-level etc)

> * In module enabled, via ``importlib.enable_lazy_imports_in_module()``.

Reviewer 2: ```suggestion
* Enabled in a specific module, via ``importlib.enable_lazy_imports_in_module()``.
```

When the flag ``-L`` is passed to the Python interpreter, a new
``sys.flags.lazy_imports`` is set to ``True``, otherwise it exists as ``False``.
This flag does not reflect whether the current status of lazy imports is enabled
or disabled (as a result of using ``importlib.set_lazy_imports()`` or
``importlib.enable_lazy_imports_in_module()``). The ``-L`` flag is propagated

> from to new Python subprocesses by using ``sys.flags.lazy_imports``, so in an

Reviewer 2: ```suggestion
to new Python subprocesses based on the value of ``sys.flags.lazy_imports``, so in an
```

analogous way, new subprocesses don't necessarily reflect the current status of
lazy imports at the time the subprocess was spawned either.

Actual current status of whether lazy imports is enabled or not at any moment
can be retrieved using ``importlib.is_lazy_imports_enabled()``, which will

> return ``True`` if lazy imports is enabled or ``False`` otherwise.

Reviewer 2: the subtleties of `-L`, programmatic enablement, and subprocesses, feel overwhelming at this point of the spec. I suggest discussing the subtleties in a later sub-heading, and keep it simple here

import occurs (say ``import foo`` or ``from foo import bar``), the key ``"foo"``
or ``"bar"`` is immediately added to the module namespace dictionary, but with
its value set to an internal-only "lazy import" object that preserves all the
necessary metadata to execute the import later. A new boolean flag in
``PyDictKeysObject`` (``dk_lazy_imports``) is set to signal that this particular
dictionary may contain lazy import objects.

> Anytime a key is looked up in a dictionary, as to extract any of its values, the

Reviewer: ```suggestion
Anytime a key is looked up in a dictionary to extract its value, the
```

value is checked to see if it is a lazy import object. If so, the lazy object is
immediately resolved, the relevant imported modules executed, the lazy import

> object is replaced in the dictionary (whenever is possible) by the actual

Reviewer: I think if we are going to add this "if possible" parenthetical, we need to add a parenthesized sentence or paragraph immediately after explaining when it might not be possible and what happens then. Otherwise it just raises unanswered questions.

```suggestion
object is replaced in the dictionary (if possible) by the actual
```

import occurs (say ``import foo`` or ``from foo import bar``), the key ``"foo"``
or ``"bar"`` is immediately added to the module namespace dictionary, but with
its value set to an internal-only "lazy import" object that preserves all the
necessary metadata to execute the import later. A new boolean flag in
``PyDictKeysObject`` (``dk_lazy_imports``) is set to signal that this particular
dictionary may contain lazy import objects.

Anytime a key is looked up in a dictionary to extract its value, the
value is checked to see if it is a lazy import object. If so, the lazy object is
immediately resolved, the relevant imported modules executed, the lazy import
object is replaced in the dictionary (if possible) by the actual imported value,

> and such resolved value is returned from the lookup function. A dictionary could

Reviewer: ```suggestion
and the resolved value is returned from the lookup function. A dictionary could
```

mutate as part of an import side effect while resolving a lazy import object.
In this case it is not possible to efficiently replace the key value with the

> resolved object. For this, the lazy import object contains a cached pointer to
> the resolved object, to be returned immediately upon next access and the attempt
> to replace the lazy import object with the resolved value will be re-tried.

Reviewer: ```suggestion
resolved object. In this case, the lazy import object will gain a cached pointer to
the resolved object. On next access that cached reference will be returned and
the lazy import object will be replaced in the dict with the resolved value.
```

Python, and at this point will replace the lazy import object at
``modb.__dict__["foo"]`` with the actual module ``foo``.

There are two cases in which a lazy import object can "escape" one dictionary:

* Into another dictionary: to preserve the performance of bulk-copy operations
  like ``dict.update()`` and ``dict.copy()``, they do not check for or resolve
  lazy import objects. However, if the source dict has the ``dk_lazy_imports``
  flag set that indicates it might contain lazy objects, that flag will be
  passed on to the updated/copied dictionary. This still ensures that the lazy
  import object can't escape into Python code without being resolved.

* Through the garbage collector: lazy imported objects are still Python objects
  and live within the garbage collector; as such, they can be collected and seen
  by means of using ``gc.collect()`` and ``gc.get_objects()``. Lazy objects are,
  however, self contained and should pose no danger if extracted from the
  garbage collector in this way.

Other "bulk" dictionary lookup methods involving values (such as ``dict.items()``,
``dict.values()``, ``PyDict_Next()`` etc.) will attempt to resolve *all* lazy

> import objects in the dictionary prior starting the lookup. Since only (some)

Reviewer: ```suggestion
import objects in the dictionary prior to starting the iteration. Since only (some)
```

module namespace dictionaries will ever have ``dk_lazy_imports`` set, the extra
overhead of resolving all lazy import objects inside a dictionary is only paid
by those dictionaries that need it. Minimizing the overhead on normal non-lazy
dictionaries is the sole purpose of the ``dk_lazy_imports`` flag.

> Given the fact that ``PyDict_Next`` could attempt to resolve all lazy import

Reviewer: (Choosing a random spot in here to leave this comment -- I think this "implementation" section of the PEP has grown large and kind of disorganized, and we should probably better organize it with some subsections, like e.g. "how a given import is determined to be lazy or not", "effect on the `PyDict` C API" (which itself should be better organized and more clearly comprehensive by listing all public C API of dicts and the effect of lazy imports on each), etc.

objects the first time position ``0`` is being accessed, imported objects could
fail with exceptions. In the cases where objects can't be resolved,

> ``PyDict_Next`` will return ``0`` ignoring any errors.

Reviewer: ```suggestion
``PyDict_Next`` will attempt to resolve all lazy import objects the first time position ``0`` is accessed, and those imports could fail with exceptions. Since ``PyDict_Next`` cannot set an exception,
``PyDict_Next`` will return ``0`` immediately in this case, and any exception will be printed to
stderr as an unraisable exception.
```

For this reason, this PEP introduces ``PyDict_NextWithError``, that works in the
same way as ``PyDict_Next``, but which can set an error when returning ``0`` and

> this should be checked via ``PyErr_Ocurred()`` after the call.

Reviewer: ```suggestion
this should be checked via ``PyErr_Occurred()`` after the call.
```

The eagerness of imports within ``try`` / ``except`` / ``with`` blocks or within
class or function bodies is handled in the compiler via a new
``EAGER_IMPORT_NAME`` opcode that always imports eagerly. Top-level imports use
``IMPORT_NAME``, which may be lazy or eager depending on ``-L`` and/or
``importlib.set_lazy_imports()``.

The current status of lazy imports at any given place is determined by a
combination of the passed ``-L`` option flag; an interpreter-wide flag set by
``importlib.set_lazy_imports()`` and the container object passed in its
``excluding`` keyword argument; and a flag set by
``importlib.enable_lazy_imports_in_module()`` in nearest running module frame.
All these together are used to cache the current status of lazy imports in the
currently running frame, this cache is globally busted whenever any of these API
functions is called and, as such, changes are immediate for all new import
statements.

Exceptions and Deferred Exceptions
----------------------------------

Exceptions that occur during a lazy import bubble up and erase the
partially-constructed module(s) from ``sys.modules``, just as exceptions during
normal import do.

> Since errors raised during a lazy import will occur later (wherever the imported

Reviewer: I think the remainder of this probably belongs better below in the "drawbacks" section (or the "backwards compatibility" section where you removed it from) instead of here. (Or maybe it should be both places, with a quick specification of the behavior here and slightly more discussion of the backwards-compatibility and debugging implications below.)

name is first referenced) than they would if the import were eager, it is
also possible that they could be accidentally caught by exception handlers that
didn't expect the import to be running within their ``try`` block, leading to
confusion.

Care should be taken when debugging unexpected exceptions keeping in mind any
and all exceptions arising from import (including ``ModuleNotFoundError``) are
deferred from import time to first-use time, which could complicate debugging.
Referencing a name in the middle of any code could trigger a deferred import and

> produce an exception while loading and executing the related imported module.

Reviewer: I don't think this paragraph adds any value over what is already said in the previous paragraph; I would just remove it. "Care should be taken" is not useful advice, and the rest of the paragraph is basically just repeating what the previous paragraph said.

Python, and at this point will replace the lazy import object at
``modb.__dict__["foo"]`` with the actual module ``foo``.

There are two cases in which a lazy import object can "escape" one dictionary:

* Into another dictionary: to preserve the performance of bulk-copy operations
  like ``dict.update()`` and ``dict.copy()``, they do not check for or resolve
  lazy import objects. However, if the source dict has the ``dk_lazy_imports``
  flag set that indicates it might contain lazy objects, that flag will be
  passed on to the updated/copied dictionary. This still ensures that the lazy
  import object can't escape into Python code without being resolved.

* Through the garbage collector: lazy imported objects are still Python objects
  and live within the garbage collector; as such, they can be collected and seen

>   by means of using ``gc.collect()`` and ``gc.get_objects()``. Lazy objects are,
>   however, self contained and should pose no danger if extracted from the
>   garbage collector in this way.

Reviewer: ```suggestion
  by means of using ``gc.collect()`` and ``gc.get_objects()``. Lazy objects are
  not useful but they are also harmless and pose no danger if extracted from the
  garbage collector in this way.
```

position ``0`` is accessed, and those imports could fail with exceptions. Since
``PyDict_Next`` cannot set an exception, ``PyDict_Next`` will return ``0``
immediately in this case, and any exception will be printed to stderr as an
unraisable exception.

For this reason, this PEP introduces ``PyDict_NextWithError``, that works in the
same way as ``PyDict_Next``, but which can set an error when returning ``0`` and
this should be checked via ``PyErr_Occurred()`` after the call.

The eagerness of imports within ``try`` / ``except`` / ``with`` blocks or within
class or function bodies is handled in the compiler via a new
``EAGER_IMPORT_NAME`` opcode that always imports eagerly. Top-level imports use
``IMPORT_NAME``, which may be lazy or eager depending on ``-L`` and/or
``importlib.set_lazy_imports()``.

The current status of lazy imports at any given place is determined by a
combination of the passed ``-L`` option flag; an interpreter-wide flag set by
``importlib.set_lazy_imports()`` and the container object passed in its
``excluding`` keyword argument; and a flag set by
``importlib.enable_lazy_imports_in_module()`` in nearest running module frame.

> All these together are used to cache the current status of lazy imports in the
> currently running frame, this cache is globally busted whenever any of these API
> functions is called and, as such, changes are immediate for all new import
> statements.

Reviewer: ```suggestion
All these together are used to cache the current status of lazy imports in the
currently running frame. This cache is globally busted whenever any of these API
functions is called so that changes take effect immediately for all new import
statements.
```

The more difficult case can occur if an import in third-party code that can't
easily be modified must be forced to be eager. For this purpose,

> ``importlib.set_lazy_imports()`` takes two optional argument: a boolean, ``True``

Reviewer: ```suggestion
``importlib.set_lazy_imports()`` takes two optional arguments: a boolean, ``True``
```

a given import will be eager or lazy.

Per-module opt-in
-----------------

Experience with the reference implementation suggests that the most practical
adoption path for lazy imports is for a specific deployed application to opt-in
globally, observe whether anything breaks, and opt-out specific modules as
needed.

It is less practical to achieve robust and significant startup-time or
memory-use wins by piecemeal application of lazy imports. Generally it would

> require blanket application of the ``__future__`` import to most of the

Reviewer: ```suggestion
require blanket application of the per-module opt-in to most of the
```

a given import will be eager or lazy.

Per-module opt-in
-----------------

Experience with the reference implementation suggests that the most practical
adoption path for lazy imports is for a specific deployed application to opt-in
globally, observe whether anything breaks, and opt-out specific modules as
needed.

It is less practical to achieve robust and significant startup-time or
memory-use wins by piecemeal application of lazy imports. Generally it would
require blanket application of the ``__future__`` import to most of the
codebase, as well as to third-party dependencies (which may be hard or
impossible.)

However, under some use cases it may be convenient to have a way to enable lazy
imports whether the application/end user requests it or not. This too can be
easily achieved::

    from importlib import enable_lazy_imports_in_module
    enable_lazy_imports_in_module()

> After calling ``enable_lazy_imports_in_module()``, every import in the module

Reviewer: I think we should explicitly reference SPEC 1 and the existing usage in scientific python libraries here, instead of being vague about who might use this.

Author: I added a line about SPEC-1 and its use case in scientific python libraries. Not sure if you feel that's enough, but flagging as resolved for now.

C API
-----

For authors of C extension modules, the proposed public C API is as follows:

> * ``importlib.set_lazy_imports(enabled: bool = True, excluding: typing.Container[str] | None = None)``

Reviewer: Is the `excluding` argument actually keyword-only? If so we should reflect that here:
```suggestion
* ``importlib.set_lazy_imports(enabled: bool = True, *, excluding: typing.Container[str] | None = None)``
```

C API
-----

For authors of C extension modules, the proposed public C API is as follows:

* ``importlib.set_lazy_imports(enabled: bool = True, excluding: typing.Container[str] | None = None)`` is available as
  ``PyObject *PyImport_SetLazyImports(PyObject *enabled, PyObject *excluding)``.

* ``importlib.is_lazy_import(dict: typing.Dict[str, object], name: str) -> bool`` is available as
  ``int PyDict_IsLazyImport(PyObject *dict, PyObject *name)``.

* ``importlib.is_lazy_imports_enabled() -> bool`` is available as
  ``int PyImport_IsLazyImportsEnabled()``.

* ``void PyDict_ResolveLazyImports(PyObject *dict)`` resolves all lazy objects in a

>   dictionary, if any. To be used prior calling ``PyDict_NextWithError()`` or

Reviewer: I don't think this sentence "to be used prior calling..." makes sense -- `PyDict_Next` now does this internally, you don't have to do it first.

Author: `PyDict_Next()` only does it for us if we're starting in position 0, otherwise it doesn't. The safest would be to always resolve the dictionary before calling; if it has already been resolved, the resolving function does nothing anyway.

Reviewer: If you want to clarify those details, that's fine, but then I would give the full context. The text that is currently there implies that you have to call it yourself without giving enough information to really understand whether/when you do or don't have to.

issue, since imports inside a ``with`` block are always eager.

> Deferred Exceptions

Reviewer: I think we still need to include this in the backward compatibility section.

Rejected Ideas
==============

Wrapping deferred exceptions
----------------------------

To reduce the potential for confusion, exceptions raised in the
course of executing a lazy import could be replaced by a ``LazyImportError``
exception (a subclass of ``ImportError``), with a ``__cause__`` set to the
original exception. The ``LazyImportError`` would have source location metadata
attached pointing the user to the original import statement, to ease
debuggability of errors from lazy imports.

Ensuring that all lazy import errors are raised as ``LazyImportError`` would
mitigate the potential confusion by reducing the likelihood that they would be
accidentally caught and mistaken for a different expected exception. However,
in practice we have seen cases, e.g. inside tests, where failing modules raise
``unittest.SkipTest`` exception and this would too end up being wrapped in
``LazyImportError``, making such tests fail because true exception types are

> being magically hidden, debunking the hypothetically likely case where
> unexpected deferred exceptions are caught by mistake as expected exceptions in
> exception handlers.

Reviewer: ```suggestion
being magically hidden. The drawbacks here seem to outweigh the hypothetical case where
unexpected deferred exceptions are caught by mistake.
```

Per-module opt-in using future imports
--------------------------------------

> A per-module opt-in using future imports (i.e.
> ``from __future__ import lazy_imports`` has one big disadvantage:
> 
> * ``__future__`` imports are not feature flags, they are for transition to

Reviewer: ```suggestion
A per-module opt-in using future imports (i.e.
``from __future__ import lazy_imports`` does not make sense because
``__future__`` imports are not feature flags, they are for transition to
```

Ensuring that all lazy import errors are raised as ``LazyImportError`` would
mitigate the potential confusion by reducing the likelihood that they would be
accidentally caught and mistaken for a different expected exception. However,
in practice we have seen cases, e.g. inside tests, where failing modules raise
``unittest.SkipTest`` exception and this would too end up being wrapped in
``LazyImportError``, making such tests fail because true exception types are
being magically hidden. The drawbacks here seem to outweigh the hypothetical case where
unexpected deferred exceptions are caught by mistake.

Per-module opt-in using future imports
--------------------------------------

A per-module opt-in using future imports (i.e.
``from __future__ import lazy_imports``) does not make sense because
``__future__`` imports are not feature flags, they are for transition to
behaviors which will become default in the future. It is not clear if lazy
imports will ever make sense as the default behavior, so we should not
promise this with a ``__future__`` import. Thus, a per-module opt-in would
require a new ``from __optional_features__ import lazy_imports`` or similar

> mechanism.

Reviewer: ```suggestion
promise this with a ``__future__`` import. Thus, the proposed  per-module opt-in
uses a function call rather than dedicated syntax. Dedicated syntax would
require a new ``from __optional_features__ import lazy_imports`` or similar
mechanism.
```
